t/129-ssl-socket.t

```diff
@@ -2528,7 +2600,318 @@ SSL reused session
 GET /t
 --- ignore_response
 --- error_log eval
-qr/\[error\] .* ngx.socket sslhandshake: expecting 1 ~ 5 arguments \(including the object\), but seen 0/
+qr/\[error\] .* ngx.socket sslhandshake: expecting 1 ~ 6 arguments \(including the object\), but seen 0/
 --- no_error_log
 [alert]
 --- timeout: 10
+
+
+
+=== TEST 33: mutual TLS handshake, upstream is not accessible without client certs
+--- http_config eval: $::mtls_http_config
+--- config eval
+"
+    server_tokens off;
+
+    location /t {
+        content_by_lua_block {
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect('unix:$::HtmlDir/mtls.sock')
+            if not ok then
+                ngx.say('failed to connect: ', err)
+            end
+
+            assert(sock:sslhandshake())
+
+            ngx.say('connected: ', ok)
+
+            local req = 'GET /\\r\\n'
+
+            local bytes, err = sock:send(req)
+            if not bytes then
+                ngx.say('failed to send request: ', err)
+                return
+            end
+
+            ngx.say('request sent: ', bytes)
+
+            ngx.say(sock:receive('*a'))
+
+            assert(sock:close())
+        }
+    }
+"
+
+--- user_files eval: $::mtls_user_files
+--- request
+GET /t
+--- response_body_like: 400 No required SSL certificate was sent
+--- no_error_log
+[alert]
+[error]
+[crit]
+[emerg]
+
+
+
+=== TEST 34: mutual TLS handshake, upstream is accessible when client certs are supplied
+--- http_config eval: $::mtls_http_config
+--- config eval
+"
+    server_tokens off;
+
+    location /t {
+        content_by_lua_block {
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect('unix:$::HtmlDir/mtls.sock')
+            if not ok then
+                ngx.say('failed to connect: ', err)
+            end
+
+            local f = assert(io.open('$::HtmlDir/mtls_client.crt'))
+            local cert_data = f:read('*a')
+            f:close()
+
+            f = assert(io.open('$::HtmlDir/mtls_client.key'))
+            local key_data = f:read('*a')
+            f:close()
+
+            local ssl = require('ngx.ssl')
+
+            local chain = assert(ssl.parse_pem_cert(cert_data))
+            local priv = assert(ssl.parse_pem_priv_key(key_data))
+
+            assert(sock:sslhandshake(nil, nil, nil, nil, { client_cert = chain, client_priv_key = priv, }))
```

Comment:
I think we should introduce a separate method instead to avoid the requirement of an extra table here. Also it makes connection pool reuse easier.

Reply:
Do you mean to implement something that is the superset of `sock:sslhandshake` just like `settimeout` and `settimeouts`? For example:

```
sock:tlshandshake({
    client_cert = chain,
    client_priv_key = priv,
    server_name = "example.com",
    ssl_verify = true,
})
```

Personally I like this design, but wasn't sure at first so went for the more conservative approach.

Reply:
Second thought. I think we can get rid of the `sock:sslhandshake` C function altogether and instead implement it (for backward compatibility) in lua-resty-core and simply calls `sock:tlshandshake`. This way we do not need to maintain two sets of logic for the same thing.